shell_test.go

```diff
@@ -407,3 +381,114 @@ func TestRefs(t *testing.T) {
 	sort.Strings(actual)
 	is.Equal(expected, actual)
 }
+
+func sleepMoment() {
+	time.Sleep(time.Millisecond * 3000)
+}
+
+func randBytes(size int) []byte {
+	b := make([]byte, size)
+	_, err := io.ReadFull(u.NewTimeSeededRand(), b)
+	if err != nil {
+		panic(err)
+	}
+	return b
+}
+
+func TestStorageUpload(t *testing.T) {
+	is := is.New(t)
+	err := utils.LoadApiConfig()
+	is.Nil(err)
+
+	s := NewShell(shellUrl)
+
+	mhash, err := s.Add(bytes.NewBufferString(string(randBytes(15))), Chunker("reed-solomon-1-1-256000"))
+	is.Nil(err)
+
+	sessionId, err := s.StorageUpload(mhash)
+	is.Nil(err)
+
+	var storage *Storage
+LOOP:
+	for {
+		storage, err = s.StorageUploadStatus(sessionId)
+		is.Nil(err)
+		switch storage.Status {
+		case "complete":
+			break LOOP
+		case "error":
+			fmt.Printf("%#v, %#v\n", storage.Status, storage.Message)
+			t.Fatal(fmt.Errorf("%s", storage.Message))
+		default:
+			fmt.Printf("%#v continue \n", storage.Status)
+			sleepMoment()
+			continue
+		}
+	}
+	fmt.Printf("Complete\n")
+}
+
+func TestStorageUploadWithOffSign(t *testing.T) {
+	is := is.New(t)
+	err := utils.LoadApiConfig()
+	is.Nil(err)
+
+	s := NewShell(shellUrl)
+
+	mhash, err := s.Add(bytes.NewBufferString(string(randBytes(15))), Chunker("reed-solomon-1-1-256000"))
+	is.Nil(err)
+
+	uts := s.GetUts()
+	sessionId, err := s.StorageUploadOffSign(mhash, uts)
+	is.Nil(err)
+
+	//var storage Storage
+LOOP:
+	for {
+		storage, err := s.StorageUploadStatus(sessionId)
+		is.Nil(err)
+		switch storage.Status {
+		case "complete":
+			break LOOP
+		case "error":
+			fmt.Printf("%#v, %#v\n", storage.Status, storage.Message)
+			t.Fatal(fmt.Errorf("%s", storage.Message))
+		case "initSignReadyEscrow", "initSignReadyGuard":
+			fmt.Printf("%#v\n", storage.Status)
+			c, err := s.StorageUploadGetContractBatch(sessionId, mhash, uts, storage.Status)
+			is.Nil(err)
+			_, err = s.StorageUploadSignBatch(sessionId, mhash, c, uts, storage.Status)
+			//is.Nil(err)
+			fmt.Printf("%#v\n", storage.Status)
+		case "balanceSignReady", "payChannelSignReady", "payRequestSignReady", "guardSignReady":
+			unsigned, err := s.StorageUploadGetUnsignedData(sessionId, mhash, uts, storage.Status)
+			is.Nil(err)
+			switch unsigned.Opcode {
+			case "balance":
+				_, err = s.StorageUploadSignBalance(sessionId, mhash, unsigned, uts, storage.Status)
+				// Note EOF returns when the daemon returns nil from the endpoint
```

Comment:
Why is that the case? Did you verify that all endpoints do this? Even so, it should be handled at a lower level, not ignoring `err` altogether

Reply:
verified error == io.EOF is returned when btfs daemon returns nil.

Reply:
I understand, but if you pass `nil` into the `Exec` for these storage upload functions, it will not trigger the io.EOF (it will behave to ignore any output), reference: https://github.com/TRON-US/go-btfs-api/blob/BTFS/requestbuilder.go#L91

Other than this the rest lgtm

Reply:
i mean from btfs daemon endpoint. I.e., for example, storageUploadSignContractBatchCmd at go-btfs, returns nil when successful (not from the go-btfs-api side). But let me double check the go-btfs-api side to see if we can avoid the io.EOF

Reply:
Not wrong to return nil from daemon. I meant skipping it on go-btfs-api side by using nil as `res` in `Exec` calls.